examples/reshare/src/dkg/state.rs

```diff
@@ -0,0 +1,608 @@
+            }
+        }
+
+        Some(Dealer::new(Some(crypto_dealer), pub_msg, unsent))
+    }
+
+    /// Create a Player for the given epoch, replaying any stored dealer messages.
+    pub fn create_player<C: Signer<PublicKey = P>>(
+        &self,
+        epoch: EpochNum,
+        signer: C,
+        round_info: Info<V, P>,
+    ) -> Option<Player<V, C>> {
+        let crypto_player =
+            CryptoPlayer::new(round_info, signer).expect("should be able to create player");
+        let mut player = Player::new(crypto_player);
+
+        // Replay persisted dealer messages (these will all be Cached responses)
+        for (dealer, pub_msg, priv_msg) in self.dealer_msgs(epoch) {
+            let _ = player.handle(dealer.clone(), pub_msg, priv_msg);
+            debug!(?epoch, ?dealer, "replayed committed dealer message");
+        }
+
+        Some(player)
+    }
+}
+
+/// Internal state for a dealer in the current round.
+pub struct Dealer<V: Variant, C: Signer> {
+    dealer: Option<CryptoDealer<V, C>>,
+    pub_msg: DealerPubMsg<V>,
+    unsent_priv_msgs: BTreeMap<C::PublicKey, DealerPrivMsg>,
+    finalized_log: Option<SignedDealerLog<V, C>>,
+}
+
+impl<V: Variant, C: Signer> Dealer<V, C> {
+    pub const fn new(
+        dealer: Option<CryptoDealer<V, C>>,
+        pub_msg: DealerPubMsg<V>,
+        unsent_priv_msgs: BTreeMap<C::PublicKey, DealerPrivMsg>,
+    ) -> Self {
+        Self {
+            dealer,
+            pub_msg,
+            unsent_priv_msgs,
+            finalized_log: None,
+        }
+    }
+
+    /// Handle an incoming ack from a player.
+    /// Returns the ack if it was successfully processed (for persistence).
+    pub fn handle(
+        &mut self,
+        player: C::PublicKey,
+        ack: PlayerAck<C::PublicKey>,
+    ) -> Option<PlayerAck<C::PublicKey>> {
+        if !self.unsent_priv_msgs.contains_key(&player) {
+            return None;
+        }
+        if let Some(ref mut dealer) = self.dealer {
+            if dealer
+                .receive_player_ack(player.clone(), ack.clone())
+                .is_ok()
+            {
+                self.unsent_priv_msgs.remove(&player);
+                return Some(ack);
+            }
+        }
+        None
+    }
+
+    /// Finalize the dealer and produce a signed log for inclusion in a block.
+    pub fn finalize(&mut self) {
+        if self.finalized_log.is_some() {
+            return;
+        }
+        if let Some(dealer) = self.dealer.take() {
+            let log = dealer.finalize();
+            self.finalized_log = Some(log);
+        }
+    }
+
+    /// Returns a clone of the finalized log if it exists.
+    pub fn finalized_log(&self) -> Option<SignedDealerLog<V, C>> {
+        self.finalized_log.clone()
+    }
+
+    /// Takes and returns the finalized log, leaving None in its place.
+    pub const fn take_finalized_log(&mut self) -> Option<SignedDealerLog<V, C>> {
+        self.finalized_log.take()
```

Comment:
A bit tricky to see that this doesn't allow duplicate finalized logs to be built because we check `self.dealer.take()` above...